File: sample-bot/bot.py
#!/usr/bin/env python3
"""Sample bot for dev-cells. Uses only Python stdlib."""

import json
import random
import sys
from http.server import HTTPServer, BaseHTTPRequestHandler

DIRECTIONS = ["up", "down", "left", "right"]

DIRECTION_DELTA = {
    "up": (0, -1),
    "down": (0, 1),
    "left": (-1, 0),
    "right": (1, 0),
}


def manhattan(x1, y1, x2, y2):
    return abs(x1 - x2) + abs(y1 - y2)


def direction_toward(cx, cy, tx, ty):
    dx = tx - cx
    dy = ty - cy
    if abs(dx) >= abs(dy):
        return "right" if dx > 0 else "left"
    return "down" if dy > 0 else "up"


def decide(cell):
    """Return an action dict for a single cell."""
    x = cell["x"]
    y = cell["y"]
    energy = cell["energy"]
    vision = cell.get("vision", {})
    food_list = vision.get("food", [])

    nearest_food = None
    best_dist = float("inf")
    for f in food_list:
        d = manhattan(x, y, f["x"], f["y"])
        if d < best_dist:
            best_dist = d
            nearest_food = f

    if energy > 30:
        if nearest_food:
            d = direction_toward(x, y, nearest_food["x"], nearest_food["y"])
        else:
            d = random.choice(DIRECTIONS)
        return {"type": "clone", "direction": d}

    if nearest_food:
        d = direction_toward(x, y, nearest_food["x"], nearest_food["y"])
        return {"type": "move", "direction": d}

    return {"type": "move", "direction": random.choice(DIRECTIONS)}


CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "POST, OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type",
}


class BotHandler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass  # suppress request logging

    def _set_cors(self):
        for k, v in CORS_HEADERS.items():
            self.send_header(k, v)

    def do_OPTIONS(self):
        self.send_response(204)
        self._set_cors()
        self.end_headers()

    def do_POST(self):
        length = int(self.headers.get("Content-Length", 0))
        body = self.rfile.read(length)
        try:
            data = json.loads(body)
        except json.JSONDecodeError:
            self.send_response(400)
            self._set_cors()
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"error": "invalid json"}).encode())
            return

        actions = {}
        for cell in data.get("cells", []):
            actions[str(cell["id"])] = decide(cell)

        response = json.dumps({"actions": actions})
        self.send_response(200)
        self._set_cors()
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(response.encode())


def main():
    port = int(sys.argv[1]) if len(sys.argv) > 1 else 3000
    server = HTTPServer(("", port), BotHandler)
    print(f"Bot running on http://localhost:{port}")
    server.serve_forever()


if __name__ == "__main__":
    main()

File: sample-bot/bot.js
#!/usr/bin/env node
/**
 * Sample bot for dev-cells. Uses only Node.js built-ins.
 * Usage: node bot.js [port]
 */

const http = require('http');

const DIRECTIONS = ['up', 'down', 'left', 'right'];

function manhattan(x1, y1, x2, y2) {
  return Math.abs(x1 - x2) + Math.abs(y1 - y2);
}

function directionToward(cx, cy, tx, ty) {
  const dx = tx - cx;
  const dy = ty - cy;
  if (Math.abs(dx) >= Math.abs(dy)) {
    return dx > 0 ? 'right' : 'left';
  }
  return dy > 0 ? 'down' : 'up';
}

function decide(cell) {
  const { x, y, energy } = cell;
  const vision = cell.vision || {};
  const food = vision.food || [];

  let nearest = null;
  let bestDist = Infinity;
  for (const f of food) {
    const d = manhattan(x, y, f.x, f.y);
    if (d < bestDist) {
      bestDist = d;
      nearest = f;
    }
  }

  if (energy > 30) {
    const dir = nearest
      ? directionToward(x, y, nearest.x, nearest.y)
      : DIRECTIONS[Math.floor(Math.random() * 4)];
    return { type: 'clone', direction: dir };
  }

  if (nearest) {
    return { type: 'move', direction: directionToward(x, y, nearest.x, nearest.y) };
  }

  return { type: 'move', direction: DIRECTIONS[Math.floor(Math.random() * 4)] };
}

const port = parseInt(process.argv[2], 10) || 3000;

const server = http.createServer((req, res) => {
  const cors = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Access-Control-Allow-Headers': 'Content-Type',
  };

  if (req.method === 'OPTIONS') {
    res.writeHead(204, cors);
    res.end();
    return;
  }

  if (req.method === 'POST') {
    let body = '';
    req.on('data', chunk => { body += chunk; });
    req.on('end', () => {
      let data;
      try {
        data = JSON.parse(body);
      } catch {
        res.writeHead(400, { ...cors, 'Content-Type': 'application/json' });
        res.end(JSON.stringify({ error: 'invalid json' }));
        return;
      }

      const actions = {};
      for (const cell of data.cells || []) {
        actions[String(cell.id)] = decide(cell);
      }

      res.writeHead(200, { ...cors, 'Content-Type': 'application/json' });
      res.end(JSON.stringify({ actions }));
    });
    return;
  }

  res.writeHead(405, cors);
  res.end();
});

server.listen(port, () => {
  console.log(`Bot running on http://localhost:${port}`);
});
